DB版メモアプリです

ham-cap opened the pull request:
DB版のメモアプリを作成いたしましたので
ご確認の程よろしくお願いいたします。

ham-cap committed:
一覧表示、新規作成、編集、削除の機能をそれぞれDBを使ったものに変更

JSONファイルとして扱っていたメモのデータを全てデータベースで管理する仕様に変更した。
SQLインジェクション対策はまだできていない。

ham-cap committed:
一部のメソッド名と変数名を変更・not_foundハンドラを追加・viewからidの表示を削除

・トップレベルに定義したmake_a_connectionメソッドが何と接続するか分かりにくかったためmake_a_connection_to_dbに変更
・patchハンドラ内のインスタンス変数@inputed_titleと@inputed_bodyはinputedなのは当たり前なので、それぞれ@new_titleと@new_bodyに変更
・存在しないURLにアクセスされた場合の対応をJSON版で入れ忘れていたためnot_foundハンドラを追加
・新規作成後や詳細画面でそのメモのidを表示していたが、特に意味はないので非表示にした

ham-cap committed:
rubocopのoffenseを修正

インデント等でoffenseが出たので修正した

ham-cap committed:
データの保存先をローカルのDBに変更・メモの更新処理を一部修正

・誰でも手元のマシンで使えるようにメモの保存先をlocalhostのDBに変更した
・メモを更新する際にedited_atカラムに値が入れ忘れていたため更新日時が入るようにした

ham-cap committed:
READMEを更新

・ローカルのDBにテーブルを設定するため、to_set_up_local_db.sqlを実行してもらうよう説明を追加

cafedomancer commented on README.md:
````diff
@@ -17,6 +17,13 @@ bundle install
+このアプリではPostgreSQLを使用します。
+PostgreSQLをインストールし、以下を実行してください。
+```
+psql postgres -f to_set_up_local_db.sql
````
たしかに postgres データベースにテーブルを作ることもできるのですが、あまり一般的な使い方ではないような気がしました。アプリケーションごとにデータベースを用意するのが一般的なので、このアプリケーション用のデータベースを作成して、それを使うようにしてみてください 🙏

cafedomancer commented on to_set_up_local_db.sql:
```diff
@@ -0,0 +1 @@
+CREATE TABLE memos(id serial PRIMARY KEY, title text NOT NULL, body text NOT NULL, created_at timestamp with time zone NOT NULL, edited_at timestamp with time zone);
```
1 行で書かれていると、このファイルを変更する必要が出てきたときに変更しづらくなってしまうような気がしました。人間がみて分かりやすいようにフォーマットしてみてください 💅

cafedomancer commented on memo_app.rb:
```diff
@@ -6,34 +6,43 @@
+  port = 5432
+  db = 'postgres'
+  user = 'postgres'
+  @connection = PG::Connection.new(host: host, port: port, dbname: db, user: user)
```
いまのコードだと `make_a_connection_to_db` メソッドを呼び出すたびにデータベースへの接続が行われそうです。データベースへの接続は重たい処理なので、いちど接続した後はそれを使い回せるようにしてみてください。

cafedomancer commented on memo_app.rb:
```diff
@@ -6,34 +6,43 @@
-  @selected_memo = @memo_files.find { |file| file['number'] == number.to_i }
+  def self.create_a_memo_array
+    make_a_connection_to_db
+    all_memo_data = @connection.exec('SELECT * FROM memos').to_a
```
現状は以下のような挙動になっていると思います。

1. `make_a_connection_to_db` メソッドを呼び出す
2. `@connection` インスタンス変数が設定される
3. `create_a_memo_to_db` メソッドで `@connection` インスタンス変数を使う

ただこれにはいくつか問題点がありそうなので対応を考えてみてください。

* 「`make_a_connection_to_db` メソッドを呼び出すと `@connection` インスタンス変数が設定され、他の場所でそれが使えるようになる」という挙動は、`make_a_connection_to_db` メソッドの中身の処理を見に行くまで挙動が分からず、メソッド名から処理を推測することができないという問題があります。メソッドの戻り値を使うようにすると、この問題を解決できると思います ✨ 

```ruby
def make_aconnection_to_db
   ...
end

connection = make_aconnection_to_db
```
 
* `@connection` は `@` 記号がついてるので、パット見は `Memo` クラスのインスタンス変数もしくはクラスインスタンス変数に見えます。しかし実体は `make_a_connection_to_db` の中で設定されるトップレベル (main オブジェクト) のインスタンス変数です。これは誤解を招く可能性があります。上記のように戻り値を使うようにすると、同じように問題を解決できると思うのでトライしてみてください。

cafedomancer replied:
メモの表示順が不定になってしまう気がしました。一貫した順序で表示できるようにしてみてください。

cafedomancer commented on memo_app.rb:
```diff
@@ -42,59 +51,43 @@ def find_selected_memo(number)
-  end
+  make_a_connection_to_db
+  @connection.prepare('create', 'INSERT INTO memos (title, body, created_at) VALUES ($1, $2, $3)')
+  @connection.exec_prepared('create', [@title, @body, @created_at])
```
一覧ページや詳細ページのメモの検索のように、メモの追加も `Memo` クラスのメソッドで行えるようにすると、より読みやすいプログラムになると思います 💡 (更新や削除についても同様)

ham-cap committed:
レビューの指摘事項を修正

・PRGパターンにするため新規作成、編集、削除の際にトップ画面にリダイレクトするように修正
・Ruby3.0以降を使用している場合、別途webrickのインストールが必要なため、Gemfileに追加
・メモの編集画面でhtmlのサニタイジングを忘れていたため修正
・使用開始までの手順をメモアプリ用のDBを作成するように変更
・to_set_up_local_db.sqlを読みやすいように改行とインデントを調整
・make_a_connection_to_dbメソッドの呼び出し回数を減らすため、返り値を定数に格納して使い回す仕様に変更
・トップページでのメモの表示順序が順不同になってしまうため、作成日時ベースで並べるよう修正
・可読性向上のため、メモの追加、更新、削除の処理をクラスメソッド化

cafedomancer commented on Gemfile:
```diff
@@ -8,5 +8,8 @@ git_source(:github) {|repo_name| "https://github.com/#{repo_name}" }
 gem "byebug"
 gem "pg"
+
+
```
余分な空行が入ってしまっているようです 👀

cafedomancer commented on Gemfile.lock:
```diff
@@ -31,8 +58,10 @@ DEPENDENCIES
   byebug
   pg
   redcarpet
+  rubocop-fjord
```
`rubocop-fjord` は `Gemfile` には記載がないので、現在の `Gemfile` の内容で `bundle install` すると `Gemfile.lock` に差分が生じてしまう気がします。差分が生じないように調整してみてください。

cafedomancer commented on to_set_up_local_db.sql:
```diff
@@ -0,0 +1,6 @@
+                    body text NOT NULL,
+                    created_at timestamp with time zone NOT NULL,
+                    edited_at timestamp with time zone
+                    );
```
スタイルが整ってないように見えてしまうのが気になりました。複数行で書くなら以下のどちらかのスタイルにするのが良いと思います。

```suggestion
CREATE TABLE memos(id serial PRIMARY KEY,
                   title text NOT NULL,
                   body text NOT NULL,
                   created_at timestamp with time zone NOT NULL,
                   edited_at timestamp with time zone);
```

```suggestion
CREATE TABLE memos(
  id serial PRIMARY KEY,
  title text NOT NULL,
  body text NOT NULL,
  created_at timestamp with time zone NOT NULL,
  edited_at timestamp with time zone
);
```

cafedomancer commented on views/top.erb:
```diff
@@ -1,7 +1,7 @@
+<% @memo_array.each do |memo| %>
+  <a href="/memos/<%= memo.id %>">
+    <p><%= "◆ #{CGI.escapeHTML(memo.title)}" %></p>
+ </a>
```
インデントがずれているようです 👀 

```suggestion
  </a>
```

cafedomancer commented on memo_app.rb:
```diff
@@ -3,37 +3,60 @@
-def sort_by_created_at
-  @memo_files = @memo_files.sort_by { |a| a['created_at'] }
-end
+Connection = make_a_connection_to_db
```
`Connection` だとクラスかモジュールに見えてしまうので、そのどちらでもないのであればすべて大文字の `CONNECTION` にしましょう 🔧

cafedomancer commented on memo_app.rb:
```diff
@@ -3,37 +3,60 @@
+
+  def self.create_a_memo_array
+    all_memo_data = Connection.exec('SELECT * FROM memos').to_a
+    all_memo_data.map { |hash| Memo.new(hash) }.sort_by(&:created_at).reverse
```
並び替えなどのデータを扱う処理は、一般的にはデータベースで行うほうがパフォーマンス的に有利なので、そうしてみてください 💁

cafedomancer commented on memo_app.rb:
```diff
@@ -3,37 +3,60 @@
+
+  def self.post_a_memo(title_params, body_params, created_at_params)
+    Connection.prepare('create', 'INSERT INTO memos (title, body, created_at) VALUES ($1, $2, $3)')
+    Connection.exec_prepared('create', [title_params, body_params, created_at_params])
```
メモの新規作成を 2 回以上行うと、以下のようなエラーが出てしまいます。確認してみてください。編集と削除についても同様です。

![localhost_4567_memos](https://user-images.githubusercontent.com/7645585/156924261-58e3b226-e40f-4023-92c4-173d2127721a.png)

cafedomancer commented on memo_app.rb:
```diff
@@ -3,37 +3,60 @@
+  end
+
+  def self.delete_a_memo(id)
+    Connection.prepare('delete', 'DELETE from memos WHERE id = $1')
```
ここだけ SQL の FROM が小文字になっているようです 👀 

```suggestion
    Connection.prepare('delete', 'DELETE FROM memos WHERE id = $1')
```

ham-cap committed:
レビューの指摘事項を修正（2）

・Gemfile.lockとのあいだに差分が生じるのを防ぐため、Gemfileにrubocop-fjordについて記載した。

・クラス名等との混同を避けるため、定数の名称をConnectionからCONNECTIONへ変更した。

・パフォーマンス向上のため、self.create_a_memo_arrayメソッドでメモデータの配列を取得する際の
並び替えをRubyではなくSQLで発行するように変更した。

・新規作成、更新、削除を2回以上行うとクエリに割り当てた名前が重複してエラーが出るため、
prepareで名前を割り当てる処理をメソッド外のトップレベルへ切り出し、メソッドではそれらを
実行するのみにした。

・SQLのFROMが小文字になっていたのを修正した。

cafedomancer commented on Gemfile:
```diff
@@ -8,5 +8,10 @@ git_source(:github) {|repo_name| "https://github.com/#{repo_name}" }
+
+inherit_gem:
+  rubocop-fjord:
+    - "config/rubocop.yml"
```
`Gemfile` に syntax error があって`bundle install` できないようです。確認してみてください 👀 

<img width="969" alt="Screen Shot 2022-03-14 at 18 44 19" src="https://user-images.githubusercontent.com/7645585/158146697-614f0da2-82d2-4b6c-8f5a-9c164373cf52.png">

ham-cap committed:
Gemfileのrubocop-fjordの記載を修正

Gemfileに書くべき内容が書けていなかったため修正

cafedomancer reviewed (approved):
🌟
